tls/s2n_tls13_secrets.c

```diff
@@ -671,3 +702,51 @@ S2N_RESULT s2n_tls13_secrets_get(struct s2n_connection *conn, s2n_extract_secret
     RESULT_ENSURE_GT(secret->size, 0);
     return S2N_RESULT_OK;
 }
+
+int s2n_connection_tls_exporter(struct s2n_connection *conn,
+        const uint8_t *label_in, uint32_t label_length,
+        const uint8_t *context, uint32_t context_length,
+        uint8_t *output_in, uint32_t output_length)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(output_in);
+    POSIX_ENSURE_REF(label_in);
+    POSIX_ENSURE_REF(context);
+    POSIX_ENSURE(s2n_connection_get_protocol_version(conn) == S2N_TLS13, S2N_ERR_INVALID_STATE);
+
+    POSIX_ENSURE(is_handshake_complete(conn), S2N_ERR_HANDSHAKE_NOT_COMPLETE);
+    POSIX_ENSURE_REF(conn->secure);
+    POSIX_ENSURE_REF(conn->secure->cipher_suite);
+    s2n_hmac_algorithm hmac_alg = conn->secure->cipher_suite->prf_alg;
+
+    struct s2n_blob label = { 0 };
+    POSIX_GUARD(s2n_blob_init(&label, (uint8_t *) label_in, label_length));
+
+    uint8_t derived_secret_bytes[S2N_TLS13_SECRET_MAX_LEN] = { 0 };
+    struct s2n_blob derived_secret = { 0 };
+    POSIX_GUARD(s2n_blob_init(&derived_secret, derived_secret_bytes,
+            s2n_get_hash_len(CONN_HMAC_ALG(conn))));
+    POSIX_GUARD_RESULT(s2n_derive_secret(hmac_alg, &CONN_SECRET(conn, exporter_master_secret),
+            &label, &EMPTY_CONTEXT(hmac_alg), &derived_secret));
+
+    DEFER_CLEANUP(struct s2n_hmac_state hmac_state = { 0 }, s2n_hmac_free);
+    POSIX_GUARD(s2n_hmac_new(&hmac_state));
+
+    DEFER_CLEANUP(struct s2n_hash_state hash = { 0 }, s2n_hash_free);
+    POSIX_GUARD(s2n_hash_new(&hash));
+
+    s2n_hash_algorithm hash_alg = { 0 };
+    POSIX_GUARD(s2n_hmac_hash_alg(hmac_alg, &hash_alg));
+    struct s2n_blob digest = EMPTY_CONTEXT(hmac_alg);
+
+    POSIX_GUARD(s2n_hash_init(&hash, hash_alg));
+    POSIX_GUARD(s2n_hash_update(&hash, context, context_length));
+    POSIX_GUARD(s2n_hash_digest(&hash, digest.data, digest.size));
```

Comment:
Wait, this doesn't look right. EMPTY_CONTEXT isn't equivalent to an zero-inited blob of the right size. It actually has a very specific value that we calculate when s2n is initialized: https://github.com/aws/s2n-tls/blob/ab7bfce9b78109c26600d1bb63c555643be1c345/tls/s2n_tls13_secrets.c#L85-L101 Aren't you overwriting that value here? That would break any future method that used EMPTY_CONTEXT.

This is another fun consequence of the blob limitations from https://github.com/aws/s2n-tls/pull/4230/files#r1343030637 And apparently EMPTY_CONTEXT needs a scary warning, or we might just want to calculate that has repeatedly at runtime to avoid this problem :(

Reply:
Fixed the immediate bug, also filed https://github.com/aws/s2n-tls/issues/4233 for follow-up.